### PR TITLE
amp-lightbox: Better resource scheduling for scrollable lightbox

### LINKED ISSUE #12685: amp-img not showing when inside a lightbox and outside of viewport initially
Start with an `<amp-list>` containing `<amp-img>` sitting inside an `<amp-lightbox>`, and with <amp-img> not initially in the screen viewport. 
When `<amp-img>` scrolls into the viewport, the image fails to lazy-load and be displayed.

Demo page: https://randomtestapp-a5a6f.firebaseapp.com/amp-img-not-showing-on-lightbox.html
Open up the lightbox, and scroll down, and you'll see empty image boxes.

### PR DESCRIPTION
**Pull request by aghassemi**
Fixes #12685

Temporary solution for high priority #12685. Layers will fix all the crazyness here.

Previous solution had the following bugs:
- Ignored new children added dynamically via amp-list
- Only monitored visibility of immediate children.
- `offsetTop` calculation were wrong and only worked when there was no other `offsetParent`
- Only layedout when child was fully visible which means user would see loading state in all case. New approach allows 2x lightboxHeight margin and prelayouts better.

(Unfortunately all the problems above also exist for scrollable version of `amp-carousel`. Given layers is the right solution and there hasn't been a high priority need, I don't recommend porting these to `amp-carousel`)

### CODE REVIEW

**Comment by aghassemi** [comment]
@erwinmombay Can we wait to get this into Canary if not cut yet?

File: extensions/amp-lightbox/0.1/amp-lightbox.js [hunk 1]
```diff
@@ -301,12 +335,21 @@ class AmpLightbox extends AMP.BaseElement {
    */
   forEachVisibleChild_(pos, callback) {
     const containerHeight = this.getSize_().height;
-    for (let i = 0; i < this.children_.length; i++) {
-      const child = this.children_[i];
-      // Check whether child element is visible in the lightbox given
+    const descendants = this.getComponentDescendants_();
+    for (let i = 0; i < descendants.length; i++) {
+      const child = descendants[i];
```
Comment (choumx) [inline extensions/amp-lightbox/0.1/amp-lightbox.js]: Nit: s/child/descendant

File: extensions/amp-lightbox/0.1/amp-lightbox.js [hunk 2]
```diff
@@ -301,12 +335,21 @@ class AmpLightbox extends AMP.BaseElement {
...
+      let offsetTop = 0;
+      for (let n = child; n && this.element.contains(n);
+        n = n./*OK*/offsetParent) {
```
Comment (choumx) [inline extensions/amp-lightbox/0.1/amp-lightbox.js]: Nit: Formatting.

```js
for (let n = child; 
    n && this.element.contains(n);
    n = n./*OK*/offsetParent) {
  offsetTop += n./*OK*/offsetTop;
}
```

File: extensions/amp-lightbox/0.1/amp-lightbox.js [hunk 3]
```diff
@@ -96,22 +123,28 @@ class AmpLightbox extends AMP.BaseElement {
...
     });
 
+    // If scrollable, take ownership of existing children and all future
+    // dynamically created children as well.
     if (this.isScrollable_) {
+      this.takeOwnershipOfDescendants_();
+
+      this.element.addEventListener(AmpEvents.DOM_UPDATE, () => {
+        this.takeOwnershipOfDescendants_();
+        this.updateChildrenInViewport_(this.pos_, this.pos_);
```
Comment (choumx) [inline extensions/amp-lightbox/0.1/amp-lightbox.js]: This means we won't call `updateInViewport` for new children. WAI?

Comment (aghassemi) [reply extensions/amp-lightbox/0.1/amp-lightbox.js]: We would, on debounced scroll we also calculate visible children and call scheduleLayout and updateInViewport on them.

**Comment by aghassemi** [comment]
@choumx Thanks. There are currently 0 tests for lightbox :( I will send a PR with some integrations tests this sprint #12759

**Pull request merged**